lib/types/hostnametrie.go

```diff
@@ -0,0 +1,162 @@
+package types
+
+import (
+	"bytes"
+	"encoding/json"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// NullHostnameTrie is a nullable HostnameTrie, in the same vein as the nullable types provided by
+// package gopkg.in/guregu/null.v3
+type NullHostnameTrie struct {
+	Trie  *HostnameTrie
+	Valid bool
+}
+
+// UnmarshalText converts text data to a valid NullHostnameTrie
+func (d *NullHostnameTrie) UnmarshalText(data []byte) error {
+	if len(data) == 0 {
+		*d = NullHostnameTrie{}
+		return nil
+	}
+	var err error
+	d.Trie, err = NewHostnameTrie(strings.Split(string(data), ","))
+	if err != nil {
+		return err
+	}
+	d.Valid = true
+	return nil
+}
+
+// UnmarshalJSON converts JSON data to a valid NullHostnameTrie
+func (d *NullHostnameTrie) UnmarshalJSON(data []byte) error {
+	if bytes.Equal(data, []byte(`null`)) {
+		d.Valid = false
+		return nil
+	}
+
+	var m []string
+	var err error
+	if err = json.Unmarshal(data, &m); err != nil {
+		return err
+	}
+	d.Trie, err = NewHostnameTrie(m)
+	if err != nil {
+		return err
+	}
+	d.Valid = true
+	return nil
+}
+
+// MarshalJSON implements json.Marshaler interface
+func (d NullHostnameTrie) MarshalJSON() ([]byte, error) {
+	if !d.Valid {
+		return []byte(`null`), nil
+	}
+	return json.Marshal(d.Trie.source)
+}
+
+// HostnameTrie is a tree-structured list of hostname matches with support
+// for wildcards exclusively at the start of the pattern. Items may only
+// be inserted and searched. Internationalized hostnames are valid.
+type HostnameTrie struct {
+	source []string
+
+	children map[rune]*HostnameTrie
+}
+
+// NewNullHostnameTrie returns a NullHostnameTrie encapsulating HostnameTrie or an error if the
+// input is incorrect
+func NewNullHostnameTrie(source []string) (NullHostnameTrie, error) {
+	h, err := NewHostnameTrie(source)
+	if err != nil {
+		return NullHostnameTrie{}, err
+	}
+	return NullHostnameTrie{
+		Valid: true,
+		Trie:  h,
+	}, nil
+}
+
+// NewHostnameTrie returns a pointer to a new HostnameTrie or an error if the input is incorrect
+func NewHostnameTrie(source []string) (*HostnameTrie, error) {
+	h := &HostnameTrie{
+		source: source,
+	}
+	for _, s := range h.source {
+		if err := h.insert(s); err != nil {
+			return nil, err
+		}
+	}
+	return h, nil
+}
+
+// Regex description of hostname pattern to enforce blocks by. Global var
+// to avoid compilation penalty at runtime.
+// based on regex from https://stackoverflow.com/a/106223/5427244
+//nolint:gochecknoglobals,lll
+var legalHostnamePattern *regexp.Regexp = regexp.MustCompile(`^(\*\.?)?((([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9]))?$`)
+
+func legalHostname(s string) error {
```

Comment:
Nit: I think `valid*` is a more suitable name here, especially since the returned error contains "invalid". I would also replace "illegal" with "invalid" in the comment on line 111.

Reply:
fixed in c4262f92